File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.java
class Solution {
    private static boolean inBound(int i , int j, int n, int m){
        return 0 <= i &&  0 <= j && i < n && j < m;
    }
    public int shortestPathBinaryMatrix(int[][] grid) {
        if (grid[0][0] == 1) return -1;
        ArrayDeque<int []> queue = new ArrayDeque();
        int n = grid.length, m = grid[0].length;
        Set<String> visited = new HashSet();
        int ans = 0;
        int [][] dir = {{-1, 0}, {1, 0}, {-1, -1}, {0, -1}, {0, 1} ,{1, -1}, {-1, 1}, {1, 1}};
        queue.addLast(new int[]{0, 0});
        visited.add("0,0");
        while (!queue.isEmpty()){
            int o = queue.size();
            ans++;
            for (int i  = 0; i < o; i++){
                int []cur = queue.pollFirst();
                if ( cur[0] == n - 1 && cur[1] == n - 1){
                    return ans;
                }
                for (int []d: dir){
                    int r = d[0] + cur[0], c = d[1] + cur[1];
                    if (inBound(r, c, n, m) && !visited.contains(r +"," + c) && grid[r][c] == 0){
                        visited.add(r + "," + c);
                        queue.addLast(new int[]{r, c});
                    }
                }
            }
        }
        return  -1;
    }
}

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        n = len(grid)
        visited = set()
        queue = deque()
        
        if grid[0][0] == 1:
            return -1
        
        queue.append((0, 0))
        visited.add((0, 0))
        directions = [(-1, 0), (1, 0), (0, 1), (0, -1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
        distance = 0
        inbound = lambda r, c: 0 <= r < n and 0 <= c < n
        
        while queue:
            length = len(queue)
            
            for i in range(length):
                x, y = queue.popleft()
                
                if x == n - 1 and y == n - 1:
                    return distance + 1
                
                for dx, dy in directions:
                    nx, ny = x + dx, y + dy
                    
                    if inbound(nx, ny) and (nx, ny) not in visited and grid[nx][ny] == 0:
                        queue.append((nx, ny))
                        visited.add((nx, ny))
                        
            distance += 1
        return -1
                
                